### PR TITLE
Make sure that macros that didn't pass LHS checking are not expanded.

### PR DESCRIPTION
**Pull request by LeoTestard**
This avoid duplicate errors for things like invalid fragment specifiers, or
parsing errors for ambiguous macros.

### CODE REVIEW

**Comment by oli-obk** [comment]
Can you add a compile-fail test for this?

**Comment by sanxiyn** [comment]
See #29231.

**Comment by LeoTestard** [comment]
Added two tests, also made sure that the help message indicating which fragment specifiers are valid is still emitted. This required to make `OnFail` able to span help messages.

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 1]
```diff
@@ -291,17 +291,16 @@ pub fn compile<'cx>(cx: &'cx mut ExtCtxt,
     let lhses = match **argument_map.get(&lhs_nm.name).unwrap() {
         MatchedSeq(ref s, _) => {
             s.iter().map(|m| match **m {
-                MatchedNonterminal(NtTT(ref tt)) => (**tt).clone(),
+                MatchedNonterminal(NtTT(ref tt)) => {
+                    valid &= check_lhs_nt_follows(cx, lhs);
```
Comment (oli-obk) [inline src/libsyntax/ext/tt/macro_rules.rs]: `lhs` should be `tt`?

Comment (LeoTestard) [inline src/libsyntax/ext/tt/macro_rules.rs]: Right, looks like I copy-pasted too fast...

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 2]
```diff
@@ -330,19 +329,19 @@ pub fn compile<'cx>(cx: &'cx mut ExtCtxt,
...
 
-fn check_lhs_nt_follows(cx: &mut ExtCtxt, lhs: &TokenTree, sp: Span) {
+fn check_lhs_nt_follows(cx: &mut ExtCtxt, lhs: &TokenTree) -> bool {
     // lhs is going to be like TokenTree::Delimited(...), where the
     // entire lhs is those tts. Or, it can be a "bare sequence", not wrapped in parens.
     match lhs {
-        &TokenTree::Delimited(_, ref tts) => {
-            check_matcher(cx, &tts.tts);
-        },
-        tt @ &TokenTree::Sequence(..) => {
```
Comment (pnkfelix) [inline src/libsyntax/ext/tt/macro_rules.rs]: I don't know if we have code formatting conventions for macro_rules but I think it looked better before (more uniform in terms of identifying the left and right hand sides of each rule )

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 3]
```diff
@@ -330,19 +329,19 @@ pub fn compile<'cx>(cx: &'cx mut ExtCtxt,
...
-        _ => cx.span_err(sp, "invalid macro matcher; matchers must be contained \
-                              in balanced delimiters or a repetition indicator")
-    };
+        &TokenTree::Delimited(_, ref tts) => check_matcher(cx, &tts.tts),
+        tt @ &TokenTree::Sequence(..) => check_matcher(cx, ref_slice(tt)),
+        _ => {
+            cx.span_err(lhs.get_span(),
+                        "invalid macro matcher; matchers must be contained \
+                         in balanced delimiters or a repetition indicator");
+            false
```
Comment (pnkfelix) [inline src/libsyntax/ext/tt/macro_rules.rs]: Okay; I assume you'll put up a separate PR that rejects repetitions as the top of the LHS, as discussed on PR #33689

File: src/test/compile-fail/macro-invalid-fragment-spec.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+macro_rules! foo(
+    ($x:foo) => () //~ ERROR invalid fragment specifier
```
Comment (pnkfelix) [inline src/test/compile-fail/macro-invalid-fragment-spec.rs]: You should include the help as an expected output

File: src/test/compile-fail/macro-missing-delimiters.rs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+macro_rules! baz(
+    baz => () //~ ERROR invalid macro matcher;
```
Comment (pnkfelix) [inline src/test/compile-fail/macro-missing-delimiters.rs]: Again include the help (HELP instead of ERROR) in the expected output

**Comment by pnkfelix** [comment]
r=me once above comments are addressed

File: src/libsyntax/ext/tt/macro_rules.rs (hunk 2 above)
Comment (LeoTestard) [inline src/libsyntax/ext/tt/macro_rules.rs]: :(

File: src/libsyntax/ext/tt/macro_rules.rs (hunk 3 above)
Comment (LeoTestard) [inline src/libsyntax/ext/tt/macro_rules.rs]: Yes, this PR's topic is to make this rejections prevent macro expansion. Not to chance those rejections themselves. :)

File: src/libsyntax/ext/tt/macro_rules.rs (hunk 2 above)
Comment (pnkfelix) [inline src/libsyntax/ext/tt/macro_rules.rs]: (To be clear I am not saying you must change if;  I don't even care enough to look up whether we have established convention here)

File: src/test/compile-fail/macro-missing-delimiters.rs (hunk 5 above)
Comment (LeoTestard) [inline src/test/compile-fail/macro-missing-delimiters.rs]: This one doesn't spawn an help message.

**Comment by LeoTestard** [comment]
@pnkfelix Fixed the test to include the help message.

File: src/test/compile-fail/macro-missing-delimiters.rs (hunk 5 above)
Comment (pnkfelix) [inline src/test/compile-fail/macro-missing-delimiters.rs]: Hmm.. shouldn't it spawn a help message?

I guess it is not important enough to block landing this PR, but I am surprised

**Comment by pnkfelix** [comment]
@LeoTestard Travis says you've put the expected error on the wrong line in your test (that,  or you need to adjust the span you use in the error reporting)

**Comment by LeoTestard** [comment]
Actually the span is fine, the error was in the test.

File: src/test/compile-fail/macro-missing-delimiters.rs (hunk 5 above)
Comment (LeoTestard) [inline src/test/compile-fail/macro-missing-delimiters.rs]: For now, the help is in the message itself: ‶invalid macro matcher; matchers must be contained blablabla″. Maybe we should split that into an error message and an help message but that's not the topic of this PR. :P
